### DISCUSSION TITLE
Max memory exceeded

### DISCUSSION THREAD

**Discussion by 108anup** [Z3Prover/z3#6556] [Q&A]
I wanted to know if there are any descriptions of when max memory exceeded error occurs.
I don't think this is a physical memory limitation.

**Context:**
I am performing many iterative solver calls. In some procedures, I try to find (binary search) a constant, until solver gives sat or unsat. In other procedures, I perform CEGIS style interactions between 2 solver instances.

In many cases, if I recreate a solver from scratch, for the same set of assertions, the solver now gives a solution instead of memory exceeded.

I am using the python API.

**Sample error**
```
File "/home/ubuntu/miniconda3/envs/wbase/lib/python3.9/site-packages/z3/z3.py", line 7084, in check
    r = Z3_solver_check_assumptions(self.ctx.ref(), self.solver, num, _assumptions)
  File "/home/ubuntu/miniconda3/envs/wbase/lib/python3.9/site-packages/z3/z3core.py", line 3968, in Z3_solver_check_assumptions
    _elems.Check(a0)
  File "/home/ubuntu/miniconda3/envs/wbase/lib/python3.9/site-packages/z3/z3core.py", line 1459, in Check
    raise self.Exception(self.get_error_message(ctx, err))
z3.z3types.Z3Exception: b'max. memory exceeded'
```

**Questions:**
1. Is there some function I can call to do cleanups instead of recreating solver. Ideally I'd want to preserve any assertions manually pushed to the solver.
2. Does reset() remove learned clauses or other solver state? I think reset() also pops any pushed assertions, is there a way to avoid that and only remove learned clauses/solver state.
3. In iterative solving, should I always call some cleanup functions before calling solvers, or should I resort to cleanups only after I retry solving on error.
4. Is there a way to change the memory limits? I think from within the python API, the limit is infinity (from solver.set()), is there a separate limit that the C implementation of the solver enforces?
5. Why is the memory limit there to begin with? What physical/logical qty does it correspond to? 

Thanks!

**Comment by NikolajBjorner** [comment]
The description of solver.reset is that it removes all asserted formulas from the solver state. It does not mean the memory is freed because there are structures, such as watch lists and pool allocators that sit around. 

3. It sounds like the solution that works better for you is to copy solver state between old solvers and new ones. Or recreate solvers from assertions. This is essentially adapted elsewhere to work around internal fragmentation accumulated by long running interactions.

5. Whatever is exposed over Pyhon is already exposed over C. You can set the memory limits at various levels: solvers, some tactics.

6. The memory limits can be configured (0 is unlimited according to your physical memory). Some systems prefer to bail out on memory pressure. See the description   https://z3prover.github.io/papers/z3internals.html#sec-a-use-case-of-tactics-from-alive2
The memory quantity is tracked based on heap allocations within z3. It tracks most (with few exceptions) heap allocations. Heap sizes are global (so not per solver).

**Reply by 108anup** [reply]
Thanks for your response!

1, 2. I am guessing z3_reset_memory() won't be useful.

3. My current strategy has been to instantiante new solver and copy the assertions from old to new. Do you think running the garbage collector explicitly help? Could you elaborate on what does "copy solver state between old and new ones" mean? Is there a way to copy over learned clauses? If we do that what would be the difference between old and new solver (in terms of memory or otherwise).

3.1. Is there a way to get assertions from different scopes (created using push operations). solver.assertions() gives all the assertions. I currently only recreate solvers in which all assertions are in the same scope. I'd want to preserve the hierarchy of assertions when creating new solver (I can bookkeep assertions myself, but the solver is already doing it somewhere, is this exposed).

4. From a max-memory-exceeded error, is there a way to know which level's memory limit was hit (solver, tactics etc.)

5. So a solver instance throwing out of memory, could mean that a different solver instance is consuming too much memory? Also I am assuming the memory limits are per process (there isn't a master z3 process that manages or enforces memory limits over all z3 processes), is this the case?
